modyn/storage/src/internal/grpc/storage_service_impl.cpp

```diff
@@ -0,0 +1,596 @@
+                                                  int64_t start_timestamp, int64_t end_timestamp) {
+  soci::session session = storage_database_connection_.get_session();
+
+  const std::vector<int64_t> file_ids = get_file_ids(dataset_id, session, start_timestamp, end_timestamp);
+
+  if (disable_multithreading_) {
+    for (const int64_t file_id : file_ids) {
+      send_samples_synchronous_retrieval<T>(writer, file_id, session);
+    }
+  } else {
+    for (const int64_t file_id : file_ids) {
+      send_samples_asynchronous_retrieval<T>(writer, file_id, session);
+    }
+  }
+}
+
+template <typename T>
+void StorageServiceImpl::send_samples_synchronous_retrieval(::grpc::ServerWriter<T>* writer, int64_t file_id,
+                                                            soci::session& session) {
+  const int64_t number_of_samples = get_number_of_samples_in_file(file_id, session);
+  if (number_of_samples > 0) {
+    soci::rowset<soci::row> rs =  // NOLINT misc-const-correctness
+        (session.prepare << "SELECT sample_id, label FROM samples WHERE file_id = :file_id", soci::use(file_id));
+    T response;
+    for (auto& row : rs) {
+      response.add_keys(row.get<long long>(0));    // NOLINT google-runtime-int
+      response.add_labels(row.get<long long>(1));  // NOLINT google-runtime-int
+      if (response.keys_size() == sample_batch_size_) {
+        writer->Write(response);
+        response.Clear();
+      }
+    }
+
+    if (response.keys_size() > 0) {
+      writer->Write(response);
+    }
+  }
+}
+
+template <typename T>
+void StorageServiceImpl::send_samples_asynchronous_retrieval(::grpc::ServerWriter<T>* writer, int64_t file_id,
+                                                             soci::session& session) {
+  const int64_t number_of_samples = get_number_of_samples_in_file(file_id, session);
+  if (number_of_samples <= sample_batch_size_) {
+    // If the number of samples is less than the sample batch size, retrieve all of the samples in one go.
+    soci::rowset<soci::row> rs =  // NOLINT misc-const-correctness
+        (session.prepare << "SELECT sample_id, label FROM samples WHERE file_id = :file_id", soci::use(file_id));
+    T response;
+    for (auto& row : rs) {
+      response.add_keys(row.get<long long>(0));    // NOLINT google-runtime-int
+      response.add_labels(row.get<long long>(1));  // NOLINT google-runtime-int
+    }
+    writer->Write(response);
+  } else {
+    // If the number of samples is greater than the sample batch size, retrieve the samples in batches of size
+    // sample_batch_size_. The batches are retrieved asynchronously and the futures are stored in a queue. When the
+    // queue is full, the first future is waited for and the response is sent to the client. This is repeated until all
+    // of the futures have been waited for.
+    std::queue<std::future<SampleData>> sample_ids_futures_queue;
+
+    for (int64_t i = 0; i < number_of_samples; i += sample_batch_size_) {
```

Comment:
I think we are mixing up two properties here: The `_sample_batch_size` is (should be) used at the _SQL Server side_. This is the maximum batch size the sql server returns to a client, at least when we use the yield_per option in Python. This is actually buffered at the sql server.

Now, we can say if we have 1 000 000 samples, and 10 threads, that each thread queries for 100000 samples. Then, if _sample_batch_size = 10000, each thread would fetch 10 partitions, but overall we would send 10 sql queries to the system. in the current implementation, we send 100 queries to the system, since each query is limited to _sample_batch_size items.

What we want (and I mentioned it in some other review comment) is to have something similar to the `yield_per` option in sqlalchemy (maybe the rowset interface is somewhat what we want but I am not sure) and then we can add the threading logic on top of that